### PR TITLE
Enable nullable reference types in analyzer projects

### PR DESCRIPTION
**Pull request by JamesNK**
Enable nullable annotations in some analyzer projects. Quite a few analyzer errors I've seen during dev and runtime are from NREs. Nullable warnings will help make our analyzer more resilient to unexpected bugs.

These are netstandard2.0 projects, so enabling nullable references is not 100% kosher, but other analyzer projects enable nullable references, e.g. https://github.com/dotnet/roslyn-analyzers, and Roslyn APIs are annotated.

### CODE REVIEW

File: src/Framework/AspNetCoreAnalyzers/src/Analyzers/Http/HeaderDictionaryAddAnalyzer.cs [hunk 1]
```diff
@@ -22,20 +23,15 @@ public override void Initialize(AnalysisContext context)
...
+        var wellKnownTypes = WellKnownTypes.GetOrCreate(context.Compilation);
 
         context.RegisterOperationAction(context =>
         {
             var invocation = (IInvocationOperation)context.Operation;
 
-            if (SymbolEqualityComparer.Default.Equals(symbols.IHeaderDictionary, invocation.Instance?.Type)
-                && IsAddMethod(invocation.TargetMethod)
-                && invocation.TargetMethod.Parameters.Length == 2)
+            if (IsAddMethod(invocation.TargetMethod)
```
Comment (JamesNK) [inline src/Framework/AspNetCoreAnalyzers/src/Analyzers/Http/HeaderDictionaryAddAnalyzer.cs:32]: Moved `IsAddMethod` first in the if statement. It's cheaper to filter out methods using that method than using SymbolEqualityComparer.

File: src/Framework/AspNetCoreAnalyzers/src/Analyzers/Microsoft.AspNetCore.App.Analyzers.csproj [hunk 2]
```diff
@@ -7,6 +7,7 @@
     <IncludeBuildOutput>false</IncludeBuildOutput>
     <Nullable>Enable</Nullable>
     <RootNamespace>Microsoft.AspNetCore.Analyzers</RootNamespace>
+    <SuppressNullableAttributesImport>true</SuppressNullableAttributesImport>
```
Comment (JamesNK) [inline src/Framework/AspNetCoreAnalyzers/src/Analyzers/Microsoft.AspNetCore.App.Analyzers.csproj:10]: Workaround https://github.com/dotnet/aspnetcore/issues/45455

File: src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteEmbeddedLanguage/FrameworkParametersCompletionProvider.cs [hunk 3]
```diff
@@ -283,23 +288,29 @@ private static bool IsArgumentTypeToken(SyntaxToken token)
         return SyntaxFacts.IsPredefinedType(token.Kind()) || token.IsKind(SyntaxKind.IdentifierToken);
     }
 
-    private static SyntaxToken? TryGetMvcActionRouteToken(CompletionContext context, SemanticModel? semanticModel, MethodDeclarationSyntax? method)
+    private static SyntaxToken? TryGetMvcActionRouteToken(CompletionContext context, SemanticModel semanticModel, MethodDeclarationSyntax? method)
```
Comment (BrennanConroy) [inline src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteEmbeddedLanguage/FrameworkParametersCompletionProvider.cs:291]:
```suggestion
    private static SyntaxToken? TryGetMvcActionRouteToken(CompletionContext context, SemanticModel semanticModel, MethodDeclarationSyntax method)
```

**Review by BrennanConroy — COMMENTED** [review]
Nice! Glad you found the root cause of nullable not working

File: src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteEmbeddedLanguage/RoutePattern/RoutePatternParser.cs [hunk 4]
```diff
@@ -172,25 +172,25 @@ private static void ValidateNoConsecutiveParameters(RoutePatternCompilationUnit
                         {
                             var message = Resources.FormatTemplateRoute_OptionalParameterHasTobeTheLast(
                                 part.Node.ToString(),
-                                previousParameterNode.GetChildNode(RoutePatternKind.ParameterName).ToString(),
-                                segmentPart.Node.ToString());
+                                previousParameterNode.GetChildNode(RoutePatternKind.ParameterName)!.ToString(),
```
[the remark this answers is not in the document]
Comment (JamesNK) [reply src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteEmbeddedLanguage/RoutePattern/RoutePatternParser.cs:175]: The API on the model is something can be a node or a token. Checking `Kind` ensures we know the content, but there is a null warning to suppress. I'll see if I can reduce them.

[the remark this answers is not in the document]
Comment (JamesNK) [reply src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteEmbeddedLanguage/RoutePattern/RoutePatternParser.cs:178]: Added an extension method that eliminates almost all bangs.

File: src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteEmbeddedLanguage/RoutePattern/RoutePatternParser.cs [hunk 5]
```diff
@@ -415,7 +417,7 @@ private RoutePatternLiteralNode ParseLiteral()
     {
         MoveBackBeforePreviousScan();
 
-        var literal = _lexer.TryScanLiteral();
+        var literal = _lexer.TryScanLiteral()!;
```
Comment (captainsafia) [inline src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteEmbeddedLanguage/RoutePattern/RoutePatternParser.cs:420]: Why can we be sure this is not null?

**Review by captainsafia — APPROVED** [review]

File: src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteEmbeddedLanguage/RoutePattern/RoutePatternParser.cs (hunk 5 above)
Comment (JamesNK) [reply src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteEmbeddedLanguage/RoutePattern/RoutePatternParser.cs:420]: The way parsing works in this specific area is that the char character has already been inspected. If it is a literal character, parsing rewinds and calls `TryScanLiteral`. A comment a few lines lower in this method explains why the result can't be null.

**Pull request merged**